### PR TITLE
bazel: improve multi-arch support for docker image artifacts

### PR DESCRIPTION
**Pull request by ixdy**
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**: yet another followup to #73930.

I discovered a few more things I'd missed:
- The hyperkube and conformance images were not made multi-arch, since by default nothing seems to depend on them. (In the make-based build, we only build them to push to gcr.io, not to gcs, and our bazel build doesn't yet push to gcr.io.)
- The "server" images (kube-proxy, kube-apiserver, etc) did not have a suffix in their docker tags.
This is because we don't use a suffix for the `-amd64` images in the make build:
https://github.com/kubernetes/kubernetes/blob/139a13d312b1b11de7f3af63d9503365ff3a1e6a/build/lib/release.sh#L361-L369
At one point in history, this was fine, since `-amd64` and `archless` tags were one and the same.
Nowadays, the archless tag actually refers to a multi-arch manifest list, so we really should be appending `-amd64` to the amd64 images.

I've also added push rules for all images; these do append the arch to the docker tags.

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```
/sig testing release

### CODE REVIEW

File: build/BUILD [hunk 1]
```diff
@@ -67,27 +67,57 @@ DOCKERIZED_BINARIES = {
     },
 }
 
-[container_image(
-    name = binary + "-internal",
+[[container_image(
```
Comment (fejta) [inline build/BUILD]: Are you not using a for loop here because that would require you to use a .bzl file? I kind of feel like that's cheating

**Review by fejta — COMMENTED** [review]
/lgtm
/hold

I feel like documenting all the alias calls would be helpful. New feature for me :)

File: build/BUILD (hunk 1 above)
Comment (ixdy) [reply build/BUILD]: correct, you can only define macros in `.bzl` files.

(I wrote this pattern enough here that maybe I should refactor it...)

**Comment by BenTheElder** [comment]
@ixdy you might appreciate what we have to do when loading images archives to ensure they match the expected tags...

```go
	// if this fails, we don't care yet, but try to get the kubernetes version
	// and see if we can skip retagging for amd64
	// if this fails, we can just assume some unknown version and re-tag
	// in a future release of kind, we can probably drop v1.11 support
	// and remove the logic below this comment entirely
	if rawVersion, err := n.KubeVersion(); err == nil {
		if ver, err := version.ParseGeneric(rawVersion); err == nil {
			if !ver.LessThan(version.MustParseSemantic("v1.12.0")) {
				return
			}
		}
	}

	// for older releases, we need the images to have the arch in their name
	// bazel built images were missing these, newer releases do not use them
	// for any builds ...
	// retag images that are missing -amd64 as image:tag -> image-amd64:tag
	// TODO(bentheelder): this is a bit gross, move this logic out of bash
	if err := n.Command(
		"/bin/bash", "-c",
		`docker images --format='{{.Repository}}:{{.Tag}}' | grep -v amd64 | xargs -L 1 -I '{}' /bin/bash -c 'docker tag "{}" "$(echo "{}" | sed s/:/-amd64:/)"'`,
	).Run(); err != nil {
		log.Warningf("Failed to re-tag docker images: %v", err)
	}
}
```

https://github.com/kubernetes-sigs/kind/blob/d31173b203f199738925303d660efb466de52072/pkg/cluster/nodes/node.go#L139-L162

**Comment by ixdy** [comment]
I added a comment for the alias() rule in `build/BUILD`.

@BenTheElder yeah, it is kinda a mess. Would it be helpful for kind if there were a filegroup for the docker image tars that contained the appropriate arch tags? 

The problem is that there's not really a good way to do this consistently. The amd64 images are also saved with no arch (in the make/bash build) because originally we had no arch suffixes at all. We added the arch suffixes to support multiarch before manifest lists were supported, but now we publish manifest lists for the arch-less names, and kubernetes manifests should just be using those, rather than the arch'd names.

This means though that a non-amd64 deployment would have different image names loaded depending on whether it pulled from a registry using the manifest list, pulled from a registry using the (largely deprected) explicit arch name, or side-loaded a docker image tarball from the server tar.

We probably should figure out a better story here, but that seems like something that maybe wants a KEP and certainly not before k8s 1.15.

I'm starting here down an approach which I think might make sense: don't ever use the arch name for the tarballs that are side-loaded. This differs from the make/bash approach, but it means that side-loading a docker image tarball gives you the same tag as pulling using the manifest list, i.e., no arch in the name at all. I added rules for the arch'd names so that in a potential follow-up we could add `container_push` rules to push to a registry. The bazel rules don't yet support manifest lists, but we could use those images to build a list.

File: build/root/BUILD.root [hunk 2]
```diff
@@ -32,6 +32,8 @@ filegroup(
     visibility = ["//visibility:private"],
 )
 
+# TODO: also add container_push rules,
```
Comment (ixdy) [inline build/root/BUILD.root]: I've basically executed this TODO locally, but I will punt to a follow-up PR

**Comment by fejta** [comment]
Can we please move logic into .bzl files? Using list comprehensions to work around the fact that you cannot have loops in BUILD.bazel files is confusing

**Review by timothysc — APPROVED** [review]
I'm not a bazel expert, but I approve of this message.  ;-)

**Comment by ixdy** [comment]
@fejta OK, this has been refactored into Starlark macros.

Because it was easy enough to add after refactoring, I also added in `docker_push` rules, so in addition to being able to build these images for multiple architectures, you can do something like
```console
$ bazel run //build:push-server-images  # push to staging-k8s.gcr.io
```
or 
```console
$ KUBE_DOCKER_PUSH_REGISTRY=gcr.io/some-custom-registry bazel run //cluster/images/conformance:push-conformance
```

**Comment by BenTheElder** [comment]
> @BenTheElder yeah, it is kinda a mess. Would it be helpful for kind if there were a filegroup for the docker image tars that contained the appropriate arch tags?

recent versions of kubernetes will not use arch tags, this is only something we do for < 1.12 IIRC, so I don't think we need that.

Ideally I just want to grab an tarball that:
- is an image containing binar{y,ies} for architecure X
- is tagged as the destination image for consumption in grc, no need for arch tags

**Comment by ixdy** [comment]
@BenTheElder I think that's what I've done here now. The tarballs intended for side-loading don't have any arch tags, but the push rules use arch tags.

/retest

File: build/BUILD [hunk 3]
```diff
@@ -31,64 +31,70 @@ filegroup(
...
     debs = select(for_platforms(
-        for_node = ["//build/debs:%s-{ARCH}.deb" % binary],
+        for_server = ["//build/debs:%s-{ARCH}.deb" % binary],
         only_os = "linux",
     )),
+    # Since the multi_arch_container macro replaces the {ARCH} format string,
+    # we need to escape the stamping vars.
+    # Also see comment above about why the push tags use ARCH while the
+    # non-push tags do not.
+    docker_push_tags = ["{{STABLE_DOCKER_PUSH_REGISTRY}}/%s-{ARCH}:{{STABLE_DOCKER_TAG}}" % binary],
```
Comment (ixdy) [inline build/BUILD]: this yields targets like `//build:push-kube-proxy`. maybe not super useful, but no harm in creating them, either.

File: hack/print-workspace-status.sh [hunk 4]
```diff
@@ -38,6 +38,8 @@ STABLE_BUILD_SCM_REVISION ${KUBE_GIT_VERSION-}
 STABLE_BUILD_MAJOR_VERSION ${KUBE_GIT_MAJOR-}
 STABLE_BUILD_MINOR_VERSION ${KUBE_GIT_MINOR-}
 STABLE_DOCKER_TAG ${KUBE_GIT_VERSION/+/_}
+STABLE_DOCKER_REGISTRY ${KUBE_DOCKER_REGISTRY:-k8s.gcr.io}
+STABLE_DOCKER_PUSH_REGISTRY ${KUBE_DOCKER_PUSH_REGISTRY:-${KUBE_DOCKER_REGISTRY:-staging-k8s.gcr.io}}
```
Comment (fejta) [inline hack/print-workspace-status.sh]: PUSH?

**Review by fejta — COMMENTED** [review]
/lgtm
/hold

File: hack/print-workspace-status.sh (hunk 4 above)
Comment (ixdy) [reply hack/print-workspace-status.sh]: I don't understand this comment.

Comment (ixdy) [reply hack/print-workspace-status.sh]: `KUBE_DOCKER_PUSH_REGISTRY` matches the pattern of `PUSH_REGISTRY` seen in other Makefiles.

**Pull request merged**